### PR TITLE
deploy: generate manifests from Helm (#482)

### LINKED ISSUE #482: Make sure helm charts don't get out of sync with our deployment manifests
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
Make sure helm charts don't get out of sync with our deployment manifests. Like https://github.com/scylladb/scylla-operator/pull/480

**What is use case behind this feature:**
When editing manifests it's easy to overlook a change that wasn't reflected in helm manifests.

**Additional Information:**
As the helm manifests are templated we can't easily use the same source. We can create a verify/update job to keep the diff between a regular and helm manifest committed as generated file. When changing a manifest without a corresponding helm edit, or the other way around, verify won't pass until both manifests are properly edited or the diff is updated.

### PR DESCRIPTION
**Pull request by zimnx**
Static manifests are generated from Helm charts.
This will guarantee that static manifests are in sync with Helm charts.

Fixes #482

### CODE REVIEW

**Review by tnozicka — DISMISSED** [review]
~/lgtm~ (sorry, had this opened from yesterday - I went only over the schema so far)

File: deploy/operator/10_certificate.yaml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 apiVersion: cert-manager.io/v1
 kind: Certificate
 metadata:
-  name: scylla-operator-serving-cert
+  name: scylla-operator-selfsigned
```
Comment (tnozicka) [inline deploy/operator/10_certificate.yaml:4]: scylla-operator-serving-cert

File: Makefile [hunk 2]
```diff
@@ -243,22 +243,67 @@ verify-crd:
 	$(diff) '$(tmp_file)' '$(CRD_PATH)' || (echo 'File $(CRD_PATH) is not up-to date. Please run `make update-crd` to update it.' && false)
 .PHONY: verify-crd
 
-# $1 - target dir
-define generate-dev-manifests
-	mkdir -p '$(1)' && \
-	cp deploy/manager/prod/*.yaml '$(1)' && \
-	yq eval '.spec.cpuset=false | .spec.datacenter.racks[] = (.spec.datacenter.racks[] | .resources.limits.cpu="200m" | .resources.limits.memory="200Mi" | .resources.requests.cpu="10m" | .resources.requests.memory="100Mi" )' -i '$(1)'/10_scylla_scyllacluster.yaml
+# $1 - name
+define generate-namespace
```
Comment (tnozicka) [inline Makefile:247]: I don't think we want to inline manifests into the makefile. If it's not coming from helm, why don't we just save the file in the target directory?

File: Makefile [hunk 3]
```diff
@@ -243,22 +243,67 @@ verify-crd:
...
-update-deploy-dev:
-	$(call generate-dev-manifests,deploy/manager/dev)
-.PHONY: update-deploy-dev
+# $1 - name
+# $2 - chart path
+# $3 - values path
+# $4 - target dir
+define generate-manifests
+	mkdir -p '$(4)'
+	$(file > $(4)/00_namespace.yaml,$(generate-namespace))
```
Comment (tnozicka) [inline Makefile:260]: `$(file >` can you elaborate on what this does? I am not familiar with such syntax in the makefile, couldn't find buildin function `file` in make reference either.

File: Makefile [hunk 4]
```diff
@@ -243,22 +243,67 @@ verify-crd:
...
 endef
 
-update-deploy-dev:
-	$(call generate-dev-manifests,deploy/manager/dev)
-.PHONY: update-deploy-dev
+# $1 - name
+# $2 - chart path
+# $3 - values path
+# $4 - target dir
+define generate-manifests
```
Comment (tnozicka) [inline Makefile:258]: generate-manifests-from-helm

File: Makefile [hunk 5]
```diff
@@ -243,22 +243,67 @@ verify-crd:
...
+
+update-deploy-manager: update-deploy-manager-dev update-deploy-manager-prod
+.PHONY: update-deploy-manager
+
+verify-deploy-manager: verify-deploy-manager-dev verify-deploy-manager-prod
+.PHONY: verify-deploy-manager
+
+update-deploy-operator:
+	$(call generate-manifests,scylla-operator,helm/scylla-operator,deploy/operator.yaml,deploy/operator)
+	ln -sf ../../$(CRD_PATH) deploy/operator/00_scylla.scylladb.com_scyllaclusters.yaml
```
Comment (tnozicka) [inline Makefile:292]: the symlink can be manual

**Review by tnozicka — COMMENTED** [review]
first pass, but I'd not generate the directory from scratch. How about creating the dir and some files like namespace or symlinks manually and just wiring updates to the other files?

File: Makefile [hunk 6]
```diff
@@ -243,22 +243,67 @@ verify-crd:
...
+# $2 - chart path
+# $3 - values path
+# $4 - target dir
+define generate-manifests
+	mkdir -p '$(4)'
+	$(file > $(4)/00_namespace.yaml,$(generate-namespace))
+	helm template '$(1)' '$(2)' --namespace '$(1)' --values '$(3)' | ./hack/split-helm.sh '$(4)'
+endef
+
+update-deploy-manager-prod:
```
Comment (tnozicka) [inline Makefile:264]: nit: Although your version is better for parallelism, I'd be fine if we merge those into just `update-deploy` for simplicity. I am fine with whichever you prefer. Just wanted to write it down so you know you have the option.

File: deploy/manager_dev.yaml [hunk 7]
```diff
@@ -0,0 +1,46 @@
+# Scylla Manager image
```
Comment (tnozicka) [inline deploy/manager_dev.yaml:1]: I'd keep the heml vars somewhere else so it doesn't confuse users under deploy/ dir which isn't really connected to helm. We may change the source in the future, split it, or not.

File: Makefile [hunk 8]
```diff
@@ -243,22 +243,67 @@ verify-crd:
...
-	$(call generate-dev-manifests,deploy/manager/dev)
-.PHONY: update-deploy-dev
+# $1 - name
+# $2 - chart path
+# $3 - values path
+# $4 - target dir
+define generate-manifests
+	mkdir -p '$(4)'
+	$(file > $(4)/00_namespace.yaml,$(generate-namespace))
+	helm template '$(1)' '$(2)' --namespace '$(1)' --values '$(3)' | ./hack/split-helm.sh '$(4)'
```
Comment (tnozicka) [inline Makefile:261]: Would it be better to pipe it into `yq` and use the structured file split?

Comment (tnozicka) [reply Makefile:261]: or `helm template --output-dir` to tmpdir and mv / rename files ?

File: Makefile (hunk 3 above)
Comment (zimnx) [reply Makefile]: no longer used, but it allows to save templatized definition in file. https://www.gnu.org/software/make/manual/html_node/File-Function.html

File: deploy/manager_dev.yaml (hunk 7 above)
Comment (zimnx) [reply deploy/manager_dev.yaml]: moved to helm/deploy/

File: Makefile (hunk 8 above)
Comment (zimnx) [reply Makefile]: i wasn't able to use yq, but i replaced awk script with Helm `--output-dir` and couple of seds and finds.

**Review by tnozicka — COMMENTED** [review]
Would this be better
https://github.com/tnozicka/scylla-operator/commit/b5a3675e862da19794bdc2c3adb1a22a8cf1a1c3
then forcing it all into runlevel 10?

File: Makefile [hunk 9]
```diff
@@ -244,22 +244,89 @@ verify-crd:
...
+
+	@leftovers=$$( find '$(3)'/scylla-manager/ -mindepth 1 -type f ) && [[ "$${leftovers}" == "" ]] || \
+	( echo -e "Internal error: Unhandled helm files: \n$${leftovers}" && false )
+endef
+
+update-deploy: tmp_dir:=$(shell mktemp -d)
+update-deploy:
+	$(call generate-operator-manifests,helm/deploy/operator.yaml,deploy/operator,$(tmp_dir))
+	$(call generate-manager-manifests,helm/deploy/manager_prod.yaml,deploy/manager/prod,$(tmp_dir))
+	$(call generate-manager-manifests,helm/deploy/manager_dev.yaml,deploy/manager/dev,$(tmp_dir))
```
Comment (tnozicka) [inline Makefile:311]: This won't verify e.g. the manual `00_namespace.yaml` file matches (or any future manual files) - would it be safer to cp `deploy/manager/prod` into  `deploy/manager/dev` and just `yq` edit the resources?

File: Makefile [hunk 10]
```diff
@@ -244,22 +244,89 @@ verify-crd:
...
+.PHONY: update-deploy
+
+verify-deploy: tmp_dir :=$(shell mktemp -d)
+verify-deploy: tmp_dir_generate :=$(shell mktemp -d)
+verify-deploy:
+	mkdir -p $(tmp_dir)/{operator,manager}
```
Comment (tnozicka) [inline Makefile:317]: nit `/{operator,manager/{prod,dev}}`

File: Makefile (hunk 9 above)
Comment (tnozicka) [reply Makefile:311]: also `helm/deploy/manager_prod.yaml` and `helm/deploy/manager_dev.yaml` could get out of sync

File: deploy/manager/prod/10_manager_configmap.yaml [hunk 11]
```diff
@@ -7,7 +7,7 @@ data:
   scylla-manager.yaml: |-
     http: :5080
     logger:
-      level: info
+      level: debug
```
Comment (tnozicka) [inline deploy/manager/prod/10_manager_configmap.yaml:10]: nit: info

File: deploy/manager/prod/50_scyllacluster.yaml [hunk 12]
```diff
@@ -4,23 +4,28 @@ metadata:
...
 spec:
-  agentVersion: 2.2.0
+  version: 4.3.2
+  agentVersion: 2.2.4
+  repository: docker.io/scylladb/scylla
+  agentRepository: docker.io/scylladb/scylla-manager-agent
+  developerMode: true
+  cpuset: true
+  automaticOrphanedNodeCleanup: false
+  network:
```
Comment (tnozicka) [inline deploy/manager/prod/50_scyllacluster.yaml:14]: nit: can we get rid of those omitempty values?

**Review by tnozicka — COMMENTED** [review]
Thanks for updating it. The last outstanding issue is handling dev manifests, otherwise this lgtm.

**Comment by tnozicka** [comment]
@zimnx would you mind picking the updates in https://github.com/tnozicka/scylla-operator/commit/9ef22bd934632ccb5b61ca486c201bf1aa95a493 and splitting them to the appropriate commits?

**Review by tnozicka — APPROVED** [review]
/lgtm
thanks!

**Pull request merged**